File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        total_product = 1
        one_zero_product = 1
        zeros = 0
        
        for i in range(len(nums)): 
            if nums[i]==0:
                zeros+=1
            elif nums[i]!=0:
                one_zero_product = one_zero_product * nums[i]
            total_product = total_product*nums[i]
        
        for i in range(len(nums)):
            if zeros>1:
                nums[i]=0
                continue
            if nums[i]==0:
                nums[i] = one_zero_product
            else:
                nums[i] = total_product//nums[i]
                
        return nums

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut zeros = 0i32;
        let mut ones_zero_prod = 1i32;
        let mut total_prod = 1i32;
        
        for i in 0..nums.len(){
            if nums[i]==0{
                zeros = zeros+1;
            }else{
                ones_zero_prod = ones_zero_prod * nums[i];
            }
            total_prod = total_prod * nums[i];
        }
        
        let mut ans = vec![0; nums.len()];
        for i in 0..nums.len(){
            if zeros>1{
                ans[i] = 0;
                continue
            }else if nums[i]==0{
                ans[i] = ones_zero_prod;
            }else{
                ans[i] = total_prod/nums[i];
            }
        }
        
        return ans;
    }
}